packages/api-server/src/db/query.ts

```diff
@@ -299,22 +302,37 @@ export class Query {
   private async queryLogsByBlockRange(
     fromBlock: HexNumber,
     toBlock: HexNumber,
-    address?: HexString | HexString[],
+    queryAddresses: HexString[],
+    queryTopics: FilterTopic[],
     lastPollId?: bigint,
     offset?: number
   ): Promise<Log[]> {
     const { MAX_QUERY_NUMBER } = getDatabaseRateLimitingConfiguration();
     const queryLastPollId = lastPollId || -1;
     const queryOffset = offset || 0;
-    let logs: DBLog[] = await this.knex<DBLog>("logs")
-      .modify(buildQueryLogAddress, address)
+
+    // NOTE: In this SQL, there is no `ORDER BY id` as combining `ORDER BY` and `LIMIT` consumes too much time when the
+    // results are large. Instead, logs are sorted outside the database:
+    // - When the number of query results exceeds $MAX_QUERY_NUMBER, an error is returned.
+    // - When the queried results are less than or equal to $MAX_QUERY_NUMBER, sorting takes only a short time
+    //
+    // NOTE: Using CTE to SELECT "logs" then JOIN "transactions" is more efficient than directly SELECT JOIN
+    let selectLogs = this.knex<DBLog>("logs")
+      .modify(buildQueryLogAddress, queryAddresses)
+      .modify(buildQueryLogTopics, queryTopics)
       .where("block_number", ">=", fromBlock)
       .where("block_number", "<=", toBlock)
       .where("id", ">", queryLastPollId.toString(10))
-      .orderBy("id", "asc")
       .offset(queryOffset)
-      .limit(MAX_QUERY_NUMBER);
-    return logs.map((log) => formatLog(log));
+      .limit(MAX_QUERY_NUMBER + 1);
+    let selectLogsJoinTransactions = this.knex<DBLog>("transactions")
+      .with("logs", selectLogs)
+      .select("logs.*", "transactions.eth_tx_hash")
+      .join("logs", { "logs.transaction_hash": "transactions.hash" });
+    let logs: DBLog[] = await selectLogsJoinTransactions;
+    return logs
+      .map((log) => formatLog(log))
```

Comment:
Is it possible to consolidate `toApiLog` and `formatLog`? This way the logs only need to be mapped once.

Reply:
It might be a good idea. However, I will not refactor it in this PR. And I think it's low-priority for us.